src/hotspot/cpu/x86/stubGenerator_x86_64.cpp

```diff
@@ -5385,6 +5385,471 @@ address generate_avx_ghash_processBlocks() {
+
+    // Write the data to the keystream array
+    // Each half of the YMM has to be written 64 bytes apart from
+    // each other in memory so the final keystream buffer holds
+    // two consecutive keystream blocks.
+    __ vextracti128(Address(result, 0), ymm_aVec, 0);
+    __ vextracti128(Address(result, 64), ymm_aVec, 1);
+    __ vextracti128(Address(result, 16), ymm_bVec, 0);
+    __ vextracti128(Address(result, 80), ymm_bVec, 1);
+    __ vextracti128(Address(result, 32), ymm_cVec, 0);
+    __ vextracti128(Address(result, 96), ymm_cVec, 1);
+    __ vextracti128(Address(result, 48), ymm_dVec, 0);
+    __ vextracti128(Address(result, 112), ymm_dVec, 1);
+
+    // This function will always write 128 bytes into the key stream buffer
+    // and that length should be returned through %rax.
+    __ mov64(rax, 128);
+
+    __ leave();
+    __ ret(0);
+    return start;
+  }
+
+  /* Add mask for 4-block ChaCha20 Block calculations */
+  address chacha20_ctradd_avx512() {
+    __ align(CodeEntryAlignment);
+    StubCodeMark mark(this, "StubRoutines", "chacha20_ctradd_avx512");
+    address start = __ pc();
+    __ emit_data64(0x0000000000000000, relocInfo::none);
+    __ emit_data64(0x0000000000000000, relocInfo::none);
+    __ emit_data64(0x0000000000000001, relocInfo::none);
+    __ emit_data64(0x0000000000000000, relocInfo::none);
+    __ emit_data64(0x0000000000000002, relocInfo::none);
+    __ emit_data64(0x0000000000000000, relocInfo::none);
+    __ emit_data64(0x0000000000000003, relocInfo::none);
+    __ emit_data64(0x0000000000000000, relocInfo::none);
+    return start;
+  }
+
+  /* Scatter mask for key stream output on AVX-512 */
+  address chacha20_scmask_avx512() {
+    __ align(CodeEntryAlignment);
+    StubCodeMark mark(this, "StubRoutines", "chacha20_scmask_avx512");
+    address start = __ pc();
+    __ emit_data64(0x0000000100000000, relocInfo::none);
+    __ emit_data64(0x0000000300000002, relocInfo::none);
+    __ emit_data64(0x0000001100000010, relocInfo::none);
+    __ emit_data64(0x0000001300000012, relocInfo::none);
+    __ emit_data64(0x0000002100000020, relocInfo::none);
+    __ emit_data64(0x0000002300000022, relocInfo::none);
+    __ emit_data64(0x0000003100000030, relocInfo::none);
+    __ emit_data64(0x0000003300000032, relocInfo::none);
+    return start;
+  }
+
+  /* Scatter mask for collating key stream output in AVX-512 */
+  address generate_chacha20Block_avx512() {
+    __ align(CodeEntryAlignment);
+    StubCodeMark mark(this, "StubRoutines", "chacha20Block");
+    address start = __ pc();
+
+    Label L_twoRounds;
+    const Register state        = c_rarg0;
+    const Register result       = c_rarg1;
+    const Register loopCounter  = rcx;
+    const KRegister writeMask = k1;
+
+    const XMMRegister zmm_aVec = xmm0;
+    const XMMRegister zmm_bVec = xmm1;
+    const XMMRegister zmm_cVec = xmm2;
+    const XMMRegister zmm_dVec = xmm3;
+    const XMMRegister zmm_scratch = xmm4;
```

Comment:
We could have 5 additional scratch registers zmm_s1 .. zmm_s5 (mapping to xmm5 ... xmm9)  to keep values read from memory into registers.

Reply:
For AVX-512 I was able to get it to work with 4 scratch registers fortunately.  For AVX and AVX2 I think the same approach can work, but since there is no lanewise bit rotation instructions that I can find I need a 5th scratch register.

For the 32-bit version it is a little more complicated as there are only 8 SIMD registers to work with.  I think even there I could simply read the state from memory for one memory-to-register add instead of doing 4, and then hold the other 128-bit state lines on 3 scratch registers.  I'm going to experiment with that a bit to see how much I can limit memory fetches to get some improvements on both 64-bit and 32-bit.